EIPS/eip-6051.md

```diff
@@ -0,0 +1,469 @@
+
+`signerPubKey`  is optional. If provided, it is assumed that the implementation has the corresponding private key and the implementation MUST sign the ephemeral public key (in the form of what is to be returned). The signature algorithm is determined by the curve part of the  `version`  parameter, that is, ECDSA for secp256k1, and Ed25519 for Curve25519. And in this situation, it should be the case that the sender trusts  `signerPubKey`, no matter how this trust is maintained. If not, next request WILL be rejected by Sender application. Also see **Security Considerations**.
+
+Implementation then MUST generate random private key `r` with cryptographic secure random number generator (CSRNG), and derives ephemeral public key `R = [r]G`. Implementation SHOULD safe keep the generated key pair `(r, R)` in a secure manner in accordance with the circumstances, and SHOULD keep it only for limited duration, but the specific duration is left to individual implementations. Implementation SHOULD be able to retrieve `r` when given back the corresponding public key `R` if within said duration.
+
+Returned value is `R`, compressed if applicable. If  `signerPubKey`  is provided, then the `signature` is appended to `R`, also hex-encoded.
+
+Alternatively, `signature` could be calculated separately, and then appended to the returned data.
+
+#### R2. Request for Sender to encapsulate the private key
+
+```javascript
+request({
+	method: 'eth_encapsulatePrivateKey',
+	params: [
+		version,
+		recipient, // public key, may be followed by its signature, see signerPubKey
+		signerPubKey,
+		oob,
+		salt,
+		account
+	],
+})
+// expected return value: S || cipher
+```
+
+`recipient`  is the return value from the call to generate ephemeral key pair, with optional `signature` appended either as returned or separately.
+
+`oob`  and  `salt`  are just byte arrays.
+
+`account`  is used to identify which private key to be encapsulated. With Ethereum it is an address. Also see **Encoding of data and messages**.
+
+If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format etc. SHALL fail the call and result in empty return and optional error logs.
+
+`signerPubKey` could have been further signed by another key pair `(trusted, trustedPubKey)`, which is trusted by the Sender application. In that case, `signerPubKey` is appended with the corresponding signature data, which SHOULD be verified against `trustedPubKey`. See **Test Cases** for further clarification.
+
+The implementation shall then proceed to retrieve the private key `sk` corresponding to  `account`, and follow the **Core Algorithms** to encrypt it.
+
+The return data is a byte array which contains first the ephemeral sender public key `S` (compressed if applicable), then `cipher` including any authentication tag, that is, `S || cipher`.
+
+#### R3. Request for Recipient to unwrap and intake the private key
```

Comment:
For R3, in your typical use case, who is the "requester" and who is the "responder", and is the communication channel between them considered "secured?

Reply:
You may think of it as a function call. All requests here (R1, R2, R3) are supposed to be called within a web page:

```javascript
window.ethereum.request){
...})
```

Wallets like MetaMask, Tally Ho  or any WalletConnect apps could be called up to process the request (I am not pretty sure about WalletConnect). A hardware/cold wallet may be working behind, connected in a secure manner.

Reply:
Regarding the security or threat model, we need to prevent a malicious party from intercepting or tampering with the Recipient public key `R`, therefore the Sender application may reject the request R2 if no signature is provided, or not provided as expected (`signerPubKey` is not signed by `trustedPubKey`, just for example). But signature is not mandatory. And the intention is to provide implementation flexibility. For example, a messaging application as Recipient may not have a proper signature, in such cases, it is expected that the private key being encapsulated is dedicated for messaging purposes.

It is possible that Sender application (under attacker's control) provides a 'fake' or 'wrong' private key in R2, but then in R3 user will find out by checking out the returned address of the intake key. A private key installed by attacker may be used to decrypt confidential information or steal funds. This is unlikely, though, as if Sender application is already under attacker's control, user funds have already been compromised.